### PR TITLE
Add coccinelle semanthic patch mechanism

### PR DESCRIPTION
**Pull request by jpereira**
It adds the new targets:
```
    coccinelle.help    - Print this
    coccinelle.clean   - Clean up the build/coccinelle
    coccinelle.diff    - Print diffs for files which would be changed by coccinelle.patch
    coccinelle.patch   - Apply Coccinelle patches to all source files in the tree.
```

### CODE REVIEW

**Comment by jpereira** [comment]
Sample: the script `scripts/coccinelle/fix_use_of_NUM_ELEMENTS.cocci` is looking for `sizeof(...)/sizeof(...[])` and replacing by `NUM_ELEMENTS()` macro.

e.g: 

```
[jpereira@sugarloaf freeradius-server.git]$ make coccinelle.report
COCCINELLE-TEST scripts/coccinelle/fix_use_of_NUM_ELEMENTS.cocci
diff -u -p a/lib/util/event.c b/lib/util/event.c
--- a/lib/util/event.c
+++ b/lib/util/event.c
@@ -622,7 +622,9 @@ static int _event_fd_delete(fr_event_fd_
 		/*
 		 *	If this fails, it's a pretty catastrophic error.
 		 */
-		count = fr_event_build_evset(evset, sizeof(evset)/sizeof(*evset), &ef->active, ef, &funcs, &ef->active);
+		count = fr_event_build_evset(evset, NUM_ELEMENTS(evset),
+					     &ef->active, ef, &funcs,
+					     &ef->active);
 		if (count > 0) {
 			int ret;

@@ -776,7 +778,7 @@ int fr_event_filter_update(fr_event_list
 		}
 	}

-	count = fr_event_build_evset(evset, sizeof(evset)/sizeof(*evset), &ef->active,
+	count = fr_event_build_evset(evset, NUM_ELEMENTS(evset), &ef->active,
 				     ef, &ef->active, &curr_active);
 	if (unlikely(count < 0)) {
 	error:
@@ -884,7 +886,9 @@ int fr_event_filter_insert(TALLOC_CTX *c
 			goto free;
 		}

-		count = fr_event_build_evset(evset, sizeof(evset)/sizeof(*evset), &ef->active, ef, funcs, &ef->active);
+		count = fr_event_build_evset(evset, NUM_ELEMENTS(evset),
+					     &ef->active, ef, funcs,
+					     &ef->active);
 		if (count < 0) goto free;
 		if (count && (unlikely(kevent(el->kq, evset, count, NULL, 0, NULL) < 0))) {
 			fr_strerror_printf("Failed inserting filters for FD %i: %s", fd, fr_syserror(errno));
@@ -910,7 +914,9 @@ int fr_event_filter_insert(TALLOC_CTX *c
 		 */
 		memcpy(&active, &ef->active, sizeof(ef->active));

-		count = fr_event_build_evset(evset, sizeof(evset)/sizeof(*evset), &ef->active, ef, funcs, &ef->active);
+		count = fr_event_build_evset(evset, NUM_ELEMENTS(evset),
+					     &ef->active, ef, funcs,
+					     &ef->active);
 		if (count < 0) {
 		error:
 			memcpy(&ef->active, &active, sizeof(ef->active));
COCCINELLE-TEST scripts/coccinelle/semantic_parentheses.cocci
[jpereira@sugarloaf freeradius-server.git]$
```

File: scripts/build/coccinelle.mk [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+.PHONY: coccinelle coccinelle.help
+coccinelle: coccinelle.help
+
+#
+#  Dump out some useful information on what Coccinelle option should be called.
+#
+coccinelle.help:
+	@echo ""
+	@echo "Make targets:"
+	@echo "    coccinelle.help    - You know!"
```
Comment (arr2036) [inline scripts/build/coccinelle.mk]: Print this help text

File: scripts/build/coccinelle.mk [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	@echo "    coccinelle.clean   - Clean up the $(OUTPUT)"
```
Comment (arr2036) [inline scripts/build/coccinelle.mk]: Unsure what this means

File: scripts/build/coccinelle.mk [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	@echo "    coccinelle.report  - Dry-run showing the changes."
```
Comment (arr2036) [inline scripts/build/coccinelle.mk]: Print diffs for files which would be changed by coccinelle.patch

File: scripts/build/coccinelle.mk [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	@echo "    coccinelle.patch   - Do the modification on the file directly."
```
Comment (arr2036) [inline scripts/build/coccinelle.mk]: Apply Coccinelle patches to all source files in the tree

File: scripts/coccinelle/fix_use_of_NUM_ELEMENTS.cocci [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// Copyright (C) 2019 Network RADIUS SAS.  Licenced under CC-by-NC 4.0.
+// Development of this scripts was sponsored by Network RADIUS SAS.
+// Author: Jorge Pereira (jpereira@freeradius.org)
+// Confidence: High
+// Comments: Fix use of NUM_ELEMENTS instead of random sizeof(T)/sizeof(E)
+// Options: --no-includes
```
Comment (arr2036) [inline scripts/coccinelle/fix_use_of_NUM_ELEMENTS.cocci]: Can the comments give at least one example of the types of transformation that will be applied?  Some inline descriptions of the syntax would be useful too.

File: scripts/coccinelle/check_vb_strvalue.cocci.disabled [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// Copyright (C) 2019 Network RADIUS SAS.  Licenced under CC-by-NC 4.0.
```
Comment (arr2036) [inline scripts/coccinelle/check_vb_strvalue.cocci.disabled]: Really not appropriate to use CC-by-NC.  Should be GPLv2 or CC-BY.

File: doc/antora/modules/developers/pages/coccinelle.adoc [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+About Coccinelle
+
+* Official site http://coccinelle.lip6.fr/
+* Howto http://coccinelle.lip6.fr/papers/tutorial.pdf
+* An Introduction to Coccinelle (video) https://www.youtube.com/watch?v=buZrNd6XkEw
+
+== Maintainer work
+
+All cocci SmPL files can be found in `scripts/coccinelle/`.
```
Comment (arr2036) [inline doc/antora/modules/developers/pages/coccinelle.adoc]: Define what SmPL means before using the acronym.

File: scripts/build/coccinelle.mk (hunk 3 above)
Comment (arr2036) [reply scripts/build/coccinelle.mk]: Rename to `coccinelle.diff`, report is ambiguous, diff is not.

File: scripts/coccinelle/semantic_parentheses.cocci [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+// Confidence: High
+// Comments: Fix semantic parentheses
+// Options: --no-includes --include-headers
+
+@@
+expression E;
+constant C;
+@@
+- !E & C
++ !(E & C)
```
Comment (arr2036) [inline scripts/coccinelle/semantic_parentheses.cocci]: Hmm, useful, but wouldn't this change the meaning of a negated expression, or does '!' mean something different here? Part of the point of this ticket was to make Coccinelle easier for other developers to use, so these examples really need to be documented better.

File: doc/antora/modules/developers/pages/coccinelle.adoc [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+== Run tests
+
+We have two different modes:
+
+=== 1. Only print the diff output.
+
+[source,shell]
+----
+$ make coccinelle.report
```
Comment (arr2036) [inline doc/antora/modules/developers/pages/coccinelle.adoc]: should be .diff now

File: scripts/coccinelle/semantic_parentheses.cocci (hunk 8 above)
Comment (jpereira) [reply scripts/coccinelle/semantic_parentheses.cocci]: image

from http://coccinelle.lip6.fr/papers/tutorial.pdf page titled `A semantic patch for the !& problem`. But, I could remove it.

**Pull request merged**